server/chat-plugins/friends.ts

```diff
@@ -0,0 +1,915 @@
+			buf += `<strong>To move over your friends to this account from a different account, `;
+			buf += `sign into that account and use </strong><code>/friend requesttransfer [new name]</code>.`;
+			return buf;
+		}
+		buf += `</h3> `;
+
+		for (const key in friends) {
+			const friendArray = categorized[key].sort();
+			if (friendArray.length === 0) continue;
+			buf += `<h4>${STATUS_TITLES[key]} (${friendArray.length})</h4>`;
+			for (const friend of friendArray) {
+				buf += `<div class="pad"><div>`;
+				buf += await this.displayFriend(friend);
+				buf += `</div></div>`;
+			}
+		}
+
+		return buf;
+	}
+	async displayFriend(userid: string) {
+		const user = Users.getExact(userid); // we want this to be exact
+		const connected = user?.connected;
+		const name = Utils.escapeHTML(user ? user.name : userid);
+		const statusType = connected ?
+			`<strong style="color:${STATUS_COLORS[user!.statusType]}">\u25C9 ${STATUS_TITLES[user!.statusType]}</strong>` :
+			'\u25CC Offline';
+		let buf = connected ?
+			`<span class="username"> <strong>${name}</strong></span><span><small> (${statusType})</small></span>` :
+			Utils.html`<i>${name}</i> <small>(${statusType})</small>`;
+		buf += `<br />`;
+
+		const oldName = this.renames.get(userid);
+		if (oldName) {
+			buf += Utils.html`<small>(recently renamed from ${oldName})</small><br />`;
+		}
+
+		const curUser = Users.get(userid); // might be an alt
+		if (user?.connected) {
+			if (user.userMessage) buf += Utils.html`Status: <i>${user.userMessage}</i><br />`;
+		} else if (curUser?.id && curUser.id !== userid) {
+			buf += `<small>On an alternate account</small><br />`;
+		}
+		const lastLogin = await this.getLastLogin(userid);
+		if (lastLogin && !connected) {
+			// THIS IS A TERRIBLE HACK BUT IT WORKS OKAY
+			const time = Chat.toTimestamp(new Date(Number(lastLogin)), {human: true});
+			buf += `Last login: ${time.split(' ').reverse().join(', on ')}`;
+			buf += ` (${Chat.toDurationString(Date.now() - lastLogin, {precision: 1})} ago)`;
+		}
+		buf = `<div class="infobox">${buf}</div>`;
+		return toLink(buf);
+	}
+	async addFriend(senderName: User | string, receiverName: string) {
+		const receiverID = toID(receiverName);
+		const senderID = toID(senderName);
+		if (!receiverID) throw new Chat.ErrorMessage(`Invalid user.`);
+		return this.query({
+			type: 'transaction', statement: 'add', data: {to: receiverID, from: senderID},
+		});
+	}
+	async removeFriend(userName: string, friendName: string) {
+		const userid = toID(userName);
+		const friendID = toID(friendName);
+		if (!friendID || !userid) throw new Chat.ErrorMessage(`Invalid usernames supplied.`);
+
+		const user = Users.getExact(userid);
+		if (user) {
+			const userFriends = await this.getFriends(user);
+			if (!userFriends.has(friendID)) throw new Chat.ErrorMessage(`You do not have '${friendID}' friended.`);
+		}
+		return this.query({
+			statement: 'delete', type: 'run', data: {userid},
+		});
```

Comment:
Delete depends on `getFriends` so, if this was the right thing to do, it would need to be in a transaction.

I don't think `delete` is doing the right thing though? Its definition is `DELETE FROM friends WHERE userid = $userid OR friend = $userid` which will delete the other persons entire friend list.

What you want to do is remove where `userid = $userid AND friend = $friend`, skip the `getFriends` check, and check if it actually removed a friend with [`changes`](https://github.com/JoshuaWise/better-sqlite3/blob/master/docs/api.md#runbindparameters---object) property to do the error message